docs/FoLiA-txt.1
.TH FoLiA-txt 1 "2023 jan 18"

.SH NAME
FoLiA-txt - Convert TXT files into FoLiA.

.SH SYNOPSIS
FoLiA-txt [options] FILE

FoLiA-txt [options] DIR

.SH DESCRIPTION

When a DIR is provided,
.B FoLiA-txt
will process all files in DIR and store the result in DIR, or in
the directory provided with
.B -O

When a FILE is provided,
.B FoLiA-txt
will process that file and store its result in the directory where FILE is
found or in the directory provided with
.B -O

All files should be in TXT XML format or BZIP2 or GZ versions thereof.
When the input file(s) are zipped, the output will be too.

.SH OPTIONS
.B -O
prefix
.RS
use 'prefix' as a directory name to store the txt-ed files in.
.RE

.B --setname
set
.RS
The name for the set in which the FoLiA nodes are created. (default FoLiA-txt-set).
.RE

.B --class
classname
.RS
The classname of the FoLiA nodes that are created. (default FoLiA-txt).
.RE

.B -t
or
.B --threads
number
.RS
Number of concurrent threads to be used by the program.
.RE

.B -v
.RS
be more verbose.
.RE

.B -V
or
.B --version
.RS
Show VERSION
.RE

.B -h
or
.B --help
.RS
Show some help
.RE

.SH BUGS
possible

.SH AUTHORS
Ko van der Sloot

Martin Reynaert

e\-mail: lamasoftware@science.ru.nl


NEWS
foliautils 0.18  22-07-2022
[Ko van der Sloot]
* FoLiA-page: only add LineBreak annotation when needed
* added more tests to make check
* adapted and fixed tests
* fixed the ugly problem of temporally disabling text checking.
* start using the "system" foliadiff
* fix declarations

[Maarten van Gompel]
* FoLiA-page: added a --nomarkup parameter to revert to the old behaviour, and an extra --nostrings parameter to omit the strings #65
* added a note for the --sent option #65
* Added some comments for the ugly disable set_checktext patch, I don't like this but it seems needed (underlying libfolia issue?) #65
* Add linebreaks and t-str to the paragraph text (currently fails text validation)
* added Dockerfile and instructions
* codemeta.json: updated according to (proposed) CLARIAH requirements (CLARIAH/clariah-plus#38)

foliautils 0.17  12-07-2021
[Ko van der Sloot]
* needs libfolia 2.9 or above
* replaced TravisCI by GitHub actions
* FoLiA-correct:
  - fixex a problem with correcting FoLia with both p and s nodes
  - added support for the FoLiA 'tag' feature
  - clearer error messages
  - fixed bugs in HEMP handling
  - better handling of Ucto's ABBREVIATION* tokens
  - fixed correctings when a word has 'space="no"'.
  - some smaller fixes
  - added more tests
* FoLiA-clean:
  - improved, using new features from libfolia 2.9
* FoLiA-2text:
  - replaced '--original' parameter by a '--correction-handling' parameter
  - implemented a --honour-tags option, to interpret tag="token" tags
  - some improvement in output-file naming
* FoLiA-abby:
  - complete reworked the code
  - added '-S' and '-C' as alternatives for '--setname' and '--classname'
  - added a --keephyphens option
  - added a --addbreaks option
  - addes option --addmetrics to optionally add positional info to the
    paragraphs
  - improved handling of '-' (Hyphen)
  - add 'font_properties', 'font_id' and 'font_style' as a feature node
  - improved handling of text with spaces at 'unexpected' locations
* all modules:
  - Code refactoring and cleaning
  - added and improved tests
  - adapted man pages

foliautils 0.16  07-01-2021
[Ko vd Sloot]
* requires libfolia 2.7 or above
* provenance data is better for a lot of modules
* added better checking on invalid NCnames in some modules.
* FoLiA-abby:
  - a lot of refactoring and additions to handle font/style information
* FoLiA-pm:
  - Notes are handled correctly now
  - fixed error in xlink attributes
* FoLiA-page:
  - more types of Page files are handled now
  - fixed annotation declarations
  - fixed offset calculation (due to change in FoLiA's opinion on those)
  - page number is added as a <br> node and in the metadata
  - added a --trusttokens option. This means that Word items in the Page file
    are added as Word's in the FoLiA, embedded in Sentences.
  - added a --norefs option to avoid adding references to the original texts
* FoLiA-correct:
  - make sure that the default is to run on 1 thread
  - added a --rebase-inputclass option

* FoLiA-alto:
  - the -t option was not always handled correctly

[Maarten van Gompel]
* FoLiA-benchmark: guard against compiler optimisation #48

foliautils 0.15  15-09-2020
[Maarten van Gompel]
* FoLiA-txt: check if a string is empty after normalisation (fix for #46)
[Ko vd Sloot]
* folia-correct: fix one-off error in hemp handling (when no hemp was found) #45
* some refactoring
  - centralized definition of XML_PARSER_OPTIONS
* bugfix in threading

foliautils 0.14  15-04-2020
[Martin Reynaert]
* updated man pages

[Ko vd Sloot]
* added man pages
* revised usage() in many modules
* the default seperator in FoLiA-stats is '_' now
* fix for: https://github.com/LanguageMachines/foliautils/issues/37
* fix for: https://github.com/LanguageMachines/foliautils/issues/41
* adapted to changes in libfolia
* many small code refactorings
* FoLiA-correct is improved a lot, allowing ngram corrections in FoLiA
* FoLiA-stats accepts a 'word_in_doc' mode now
* FoLiA-alto by default ceated <w> nodes now. use --oldstring to get <str>
* improved a lot in tests/
* many small fixes

foliautils 0.13  30-08-2019
[Ko vd Sloot]
bug fixes:
* fix for https://github.com/LanguageMachines/foliautils/issues/35
* fix for https://github.com/LanguageMachines/foliautils/issues/36

[Maarten van Gompel]
new features:
* FoLiA-wordtranslate.cxx: frog should be able to deal with
  spaces now, no need for ugly non-breaking space hack that now causes
  other problems in frog (LanguageMachines/frog#77)
* FoLiA-wordtranslate.cxx: adding ability to constrain by language


foliautils 0.12  29-05-2019
[Ko vd Sloot]
Release for FoLiA 2.0.

foliautils 0.11  15-05-2019
[Ko vd Sloot]
Stabilizing release for pre FoLiA 2.0

* Updated and added some tests
* started moving common code to a separate file and build a library
  (libfoliautils)
  - hemp detection is one of them

* FoLiA-stats:
  - added possibility to read a list of directories + file-names to process
    into seperate output directories. (could be generalized to other programs)
  - better hemp detection
* FoLiA-correct:
  - use same hempo detections as FoLiA-stats
* FoLiA-abby:
  - support more flavors
* FoLiA-clean:
  - avoid removing the last remaining tekt on <w> nodes
  - cleaning of tokenization now works


foliautils 0.10  29-11-2018
[Ko vd Sloot]
  * fixed icu:namespace issues
  * added FoLiA-abby, an ABBY to FoLiA convertor
  * src/FoLiA-abby.cxx, src/FoLiA-page.cxx, src/FoLiA-pm.cxx:
     - Allow 'none' value for --prefix
  * src/FoLiA-page.cxx, src/FoLiA-hocr.cxx: fixed Alignment info
  * src/FoLiA-correct.cxx:
     - fixed a problem with correction of the last word of a trigram.
     - fix correction of paragraphs with only deeper text
     - The --rank option accepts more flavors of files
  * src/FoLiA-stats.cxx:
     - added a --detokenize option
  * several minor fixes, refactorings etc.
  * updated tests

foliautils 0.9.2  05-06-2018
[Ko vd Sloot]
Bug fix release:

src/FoLiA-txt.cxx
/*
  Copyright (c) 2014 - 2023
  CLST  - Radboud University
  ILK   - Tilburg University

  This file is part of foliautils

  foliautils is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 3 of the License, or
  (at your option) any later version.

  foliautils is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, see <http://www.gnu.org/licenses/>.

  For questions and suggestions, see:
      https://github.com/LanguageMachines/foliautils/issues
  or send mail to:
      lamasoftware (at ) science.ru.nl
*/

#include <string>
#include <iostream>
#include <fstream>

#include "ticcutils/CommandLine.h"
#include "ticcutils/StringOps.h"
#include "ticcutils/FileUtils.h"
#include "ticcutils/StringOps.h"
#include "libfolia/folia.h"
#include "foliautils/common_code.h"

#include "config.h"
#ifdef HAVE_OPENMP
#include "omp.h"
#endif

using namespace	std;
using namespace	folia;

string setname = "FoLiA-txt-set";
string classname = "FoLiA-txt";
string processor_id;

void usage(){
  cerr << "Usage: [options] file/dir" << endl;
  cerr << "\t FoLiA-txt will produce FoLiA files from text files " << endl;
  cerr << "\t The output will only contain <p> and <str> nodes." << endl;
  cerr << "\t-t <threads>\n\t--threads <threads> Number of threads to run on." << endl;
  cerr << "\t\t\t If 'threads' has the value \"max\", the number of threads is set to a" << endl;
  cerr << "\t\t\t reasonable value. (OMP_NUM_TREADS - 2)" << endl;
  cerr << "\t-h or --help\t this message" << endl;
  cerr << "\t-V or --version\t show version " << endl;
  cerr << "\t-O\t output directory " << endl;
  cerr << "\t--setname The FoLiA setname of the created nodes. "
    "(Default '" << setname << "')" << endl;
  cerr << "\t--class The classname of the <t> nodes that are created. (Default '"
       << classname << "')"<< endl;
}

inline bool is_real_empty( const UnicodeString& str ){
  return is_norm_empty( TiCC::UnicodeToUTF8(str) );
}

inline UnicodeString& pop_back( UnicodeString& us ){
  return us.remove( us.length() - 1 );
}

int main( int argc, char *argv[] ){
  TiCC::CL_Options opts( "hVt:O:", "class:,setname:,help,version,threads:" );
  try {
    opts.init( argc, argv );
  }
  catch ( exception&e ){
    cerr << e.what() << endl;
    exit(EXIT_FAILURE);
  }
  string outputDir;
  string value;
  if ( opts.empty() ){
    usage();
    exit(EXIT_FAILURE);
  }
  if ( opts.extract( 'h' ) ||
       opts.extract( "help" ) ){
    usage();
    exit(EXIT_SUCCESS);
  }
  if ( opts.extract( 'V' ) ||
       opts.extract( "version" ) ){
    cerr << PACKAGE_STRING << endl;
    exit(EXIT_SUCCESS);
  }
  string command = "FoLiA-txt " + opts.toString();
  if ( opts.extract( 't', value )
       || opts.extract( "threads", value ) ){
#ifdef HAVE_OPENMP
    int numThreads = 1;
    if ( TiCC::lowercase(value) == "max" ){
      numThreads = omp_get_max_threads() - 2;
    }
    else if ( !TiCC::stringTo(value,numThreads) ) {
      cerr << "illegal value for -t (" << value << ")" << endl;
      exit( EXIT_FAILURE );
    }
    omp_set_num_threads( numThreads );
#else
    cerr << "-t option does not work, no OpenMP support in your compiler?" << endl;
    exit( EXIT_FAILURE );
#endif
  }

  opts.extract( 'O', outputDir );
  if ( !outputDir.empty() && outputDir[outputDir.length()-1] != '/' ){
    outputDir += "/";
  }
  opts.extract( "class", classname );
  opts.extract( "setname", setname );
  if ( !outputDir.empty() ){
    if ( !TiCC::isDir(outputDir) ){
      if ( !TiCC::createPath( outputDir ) ){
	cerr << "outputdir '" << outputDir
	     << "' doesn't exist and can't be created" << endl;
	exit(EXIT_FAILURE);
      }
    }
  }
  vector<string> file_names = opts.getMassOpts();
  size_t to_do = file_names.size();
  if ( to_do == 0 ){
    cerr << "no matching files found" << endl;
    exit(EXIT_SUCCESS);
  }
  if ( to_do == 1 ){
    if ( TiCC::isDir( file_names[0] ) ){
      file_names = TiCC::searchFiles( file_names[0] );
      to_do = file_names.size();
      if ( to_do == 0 ){
	cerr << "no files found in inputdir" << endl;
	exit(EXIT_SUCCESS);
      }
    }
  }
  if ( to_do > 1 ){
    cout << "start processing of " << to_do << " files " << endl;
  }
  size_t failed_docs = 0;
#pragma omp parallel for shared(file_names) schedule(dynamic)
  for ( size_t fn=0; fn < file_names.size(); ++fn ){
    string fileName = file_names[fn];
    ifstream is( fileName );
    if ( !is ){
#pragma omp critical
      {
	cerr << "failed to read " << fileName << endl;
      }
      continue;
    }
    string nameNoExt = fileName;
    string::size_type pos = fileName.rfind( "." );
    if ( pos != string::npos ){
      nameNoExt = fileName.substr(0, pos );
    }
    string docid = nameNoExt;
    pos = docid.rfind( "/" );
    if ( pos != string::npos ){
      docid = docid.substr( pos+1 );
    }
    if ( !outputDir.empty() ){
      nameNoExt = docid;
    }
    if ( !isNCName( docid ) ){
      docid = "doc-" + docid;
      if ( !isNCName( docid ) ){
	throw ( "unable to generate a Document ID from the filename: '"
		+ fileName + "'" );
      }
    }
    Document *d = 0;
    try {
      d = new Document( "xml:id='"+ docid + "'" );
    }
    catch ( exception& e ){
#pragma omp critical
      {
	cerr << "failed to create a document with id:'" << docid << "'" << endl;
	cerr << "reason: " << e.what() << endl;
	++failed_docs;
	--to_do;
      }
      continue;
    }
    processor *proc = add_provenance( *d, "FoLiA-txt", command );
    processor_id = proc->id();
    KWargs args;
    args["processor"] = processor_id;
    d->declare( folia::AnnotationType::STRING, setname, args );
    args.clear();
    args["xml:id"] = docid + ".text";
    folia::Text *text = d->create_root<folia::Text>( args );
    int parCount = 0;
    int wrdCnt = 0;
    folia::FoliaElement *par = 0;
    string parId;
    vector<FoliaElement*> par_stack; // temp store for textfragments which will
    // make up the paragraph text. may include formatting like <t-hbr/>
    UnicodeString line;
    while ( TiCC::getline( is, line ) ){
      line.trim();
      if ( line.isEmpty() ){
	// end a paragraph
	if ( par && !par_stack.empty() ){
	  // do we have some fragments?
	  folia::KWargs text_args;
	  text_args["class"] = classname;
	  FoliaElement *txt = par->add_child<folia::TextContent>( text_args );
	  for ( const auto& it : par_stack ){
	    txt->append(it );
	  }
	  par_stack.clear();
	}
	par = 0;
	continue;
      }
      vector<UnicodeString> words = TiCC::split( line );
      for ( const auto& w : words ){
	if ( par == 0 ){
	  // start a new Paragraph, only when at least 1 entry.
	  folia::KWargs p_args;
	  p_args["processor"] = processor_id;
	  d->declare( folia::AnnotationType::PARAGRAPH, setname, p_args );
	  p_args.clear();
	  parId = docid + ".p." +  TiCC::toString(++parCount);
	  p_args["xml:id"] = parId;
	  par = text->add_child<folia::Paragraph>( p_args );
	  wrdCnt = 0;
	}
	UnicodeString str_content = w; // the value to create a String node
	str_content.trim();
	if ( !is_real_empty(str_content) ){
	  UnicodeString par_content = str_content; // the value we will use for
	  // the paragraph text
	  UnicodeString hyp; // hyphen symbol at the end of par_contet
	  if ( par_content.endsWith( "¬" ) ){
	    par_content = pop_back( par_content ); // remove it
	    hyp = "¬";  // the Not-Sign u00ac
	  }
	  else if ( par_content.endsWith( "- " ) ){
	    par_content = pop_back( par_content ); // remove the space
	    par_content = pop_back( par_content ); // remove the '-'
	    hyp = "-";
	  }
	  else if ( par_content.endsWith( "-" ) ){
	    par_content = pop_back( par_content ); // remove the '-'
	    hyp = "-";
	  }
	  // now we can add the <String>
	  folia::KWargs str_args;
	  str_args["xml:id"] = parId + ".str." +  TiCC::toString(++wrdCnt);
	  folia::FoliaElement *str = par->add_child<folia::String>( str_args );
	  str->setutext( str_content, classname );
	  if ( hyp.isEmpty() && &w != &words.back() ){
	    par_content += " "; // no hyphen, so add a space separator except
	    // for last word
	  }
	  XmlText *e = new folia::XmlText(); // create partial text
	  e->setvalue( TiCC::UnicodeToUTF8(par_content) );
	  par_stack.push_back( e ); // add the XmlText to te stack
	  if ( !hyp.isEmpty() ){
	    // add an extra HyphBreak to the stack
	    folia::KWargs args;
	    args["class"] = TiCC::UnicodeToUTF8(hyp);
	    FoliaElement *e = new folia::Hyphbreak(args,d);
	    par_stack.push_back( e );
	  }
	}
      }
    }
    if ( par && !par_stack.empty() ){
      // leftovers
      folia::KWargs text_args;
      text_args["class"] = classname;
      FoliaElement *txt = par->add_child<folia::TextContent>( text_args );
      for ( const auto& it : par_stack ){
	txt->append(it );
      }
      par_stack.clear();
    }
    if ( parCount == 0 ){
#pragma omp critical
      {
	cerr << "no useful data found in document:'" << docid << "'" << endl;
	cerr << "skipped!" << endl;
	++failed_docs;
	--to_do;
      }
      continue;
    }
    string outname = outputDir + nameNoExt + ".folia.xml";
    d->save( outname );
#pragma omp critical
    {
      cout << "Processed: " << fileName << " into " << outname
	   << " still " << --to_do << " files to go." << endl;
    }
    delete d;
  }
  if ( failed_docs > 0 && failed_docs == to_do ){
    cerr << "No documents could be handled successfully!" << endl;
    return EXIT_SUCCESS;
  }
  else {
    return EXIT_SUCCESS;
  }
}
